### PR TITLE
Adapting iperf3 to get data easily

### PR DESCRIPTION
**Pull request by RkShaRkz**
This PR is the first in line of the few necessary to implement the whole "perform and send user experience data while not moving" feature.

This focuses on making a new Iperf3 monitor which:
- monitors iperf3 results, 
- rechews (reparses) the last successful iperf3 result 
- sends the chewed data back to the listener that told him to start listening
- continues listening until shutdown()

Since it's kinda getting late, the repo with the ringbuffer will land in this or a new PR tomorrow when I wake up.

## Release Notes

- **New Features**
	- Introduced a new button to prefill input fields with sample data for network performance tests.
	- Added functionality to monitor and log iPerf3 network performance metrics.
	- New class `MetricsPOJO` for structured representation of download and upload metrics.

- **Bug Fixes**
	- Improved error handling and logging in the iPerf3 parser.

- **Documentation**
	- Enhanced logging for better traceability during execution.

- **Chores**
	- Updated access modifiers for improved visibility of certain methods.

### CODE REVIEW

**Comment by RkShaRkz** [comment]
@coderabbitai review

File: app/src/main/java/cloudcity/CloudCityUtil.java [hunk 1]
```diff
@@ -1,9 +1,26 @@
...
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.Iperf3.Iperf3ResultsDataBase;
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.Iperf3.JSON.Error;
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.Iperf3.JSON.Interval.Interval;
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.Iperf3.JSON.Interval.Sum.SUM_TYPE;
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.Iperf3.JSON.Interval.Sum.Sum;
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.Iperf3.JSON.Interval.Sum.UDP.UDP_DL_SUM;
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.Metric.METRIC_TYPE;
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.Metric.Metric;
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.R;
+
```
Comment (RkShaRkz) [inline app/src/main/java/cloudcity/CloudCityUtil.java:23]: none of this seems necessary. remove

File: app/src/main/java/cloudcity/dataholders/MetricsPOJO.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+//                ", DLmedian=" + DLmedian +
+//                ", DLmean=" + DLmean +
+//                ", DLmax=" + DLmax +
+//                ", DLlast=" + DLlast +
+//                ", ULmin=" + ULmin +
+//                ", ULmedian=" + ULmedian +
+//                ", ULmean=" + ULmean +
+//                ", ULmax=" + ULmax +
+//                ", ULlast=" + ULlast +
+//                '}';
```
Comment (RkShaRkz) [inline app/src/main/java/cloudcity/dataholders/MetricsPOJO.java:62]: can get rid. can leave it in. no need to leave it in tho.

delete this garbage.

File: app/src/main/java/de/fraunhofer/fokus/OpenMobileNetworkToolkit/Iperf3/Iperf3Fragment.java [hunk 3]
```diff
@@ -285,6 +288,24 @@ public void afterTextChanged(Editable s) {
...
+            int validPortMax = 9240;
+            Random rnd = new Random();
+            int randomPort = rnd.nextInt((validPortMax - validPortMin) + 1) + validPortMin;
+            iperf3EtPort.setText(String.valueOf(randomPort));
+
+            // Set duration to 30
+            iperf3EtDuration.setText(String.valueOf(30));
+
+            // tick of checkboxes
+            iperf3BiDir.setChecked(true);
```
Comment (RkShaRkz) [inline app/src/main/java/de/fraunhofer/fokus/OpenMobileNetworkToolkit/Iperf3/Iperf3Fragment.java:306]: some holder for this data would have been better but it does the trick for now. leave it like this.

File: app/src/main/java/cloudcity/Iperf3Monitor.java [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+                    // This is useful for when we run the test and open it up in the default Iperf3 log viewer
+                    // since the Iperf3Parser might be destructively parsing the file - from what I noticed only
+                    // one of them reaches the END
+                    String originalFilePath = latestIperf3RunResult.input.iperf3rawIperf3file;
+                    String targetFilePath = originalFilePath + ".tmp";
+
+                    try {
+                        copyFile(originalFilePath, targetFilePath);
+                    } catch (IOException e) {
+//                        throw new RuntimeException(e);
```
Comment (RkShaRkz) [inline app/src/main/java/cloudcity/Iperf3Monitor.java:146]: eh. remove.

File: app/src/main/java/cloudcity/Iperf3Monitor.java [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+                                    double DLmedian = defaultReverseThroughput.calcMedian() / 1e+6;
+                                    double DLmax = defaultReverseThroughput.calcMax() / 1e+6;
+                                    double DLmean = defaultReverseThroughput.calcMean() / 1e+6;
+                                    double DLlast = getLast(defaultReverseThroughput.getMeanList()) / 1e+6;
+
+                                    double ULmin = defaultThroughput.calcMin() / 1e+6;
+                                    double ULmedian = defaultThroughput.calcMedian() / 1e+6;
+                                    double ULmax = defaultThroughput.calcMax() / 1e+6;
+                                    double ULmean = defaultThroughput.calcMean() / 1e+6;
+                                    double ULlast = getLast(defaultThroughput.getMeanList()) / 1e+6;
```
Comment (RkShaRkz) [inline app/src/main/java/cloudcity/Iperf3Monitor.java:222]: consider extracting all this dividing into a 
```
public double normalize(double value) {
    return value / 1e+6;
}
```
method... dunno why I didn't want to do it immediatelly, but i remember not wanting to do it for some reason. can't remember which, so I was wrong.

File: app/src/main/java/cloudcity/Iperf3Monitor.java [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+
+                            double DLmin = defaultReverseThroughput.calcMin() / 1e+6;
+                            double DLmedian = defaultReverseThroughput.calcMedian() / 1e+6;
+                            double DLmax = defaultReverseThroughput.calcMax() / 1e+6;
+                            double DLmean = defaultReverseThroughput.calcMean() / 1e+6;
+
+                            double ULmin = defaultThroughput.calcMin() / 1e+6;
+                            double ULmedian = defaultThroughput.calcMedian() / 1e+6;
+                            double ULmax = defaultThroughput.calcMax() / 1e+6;
+                            double ULmean = defaultThroughput.calcMean() / 1e+6;
```
Comment (RkShaRkz) [inline app/src/main/java/cloudcity/Iperf3Monitor.java:271]: DRY violation right there :D

File: app/src/main/java/cloudcity/Iperf3Monitor.java [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+
+                            Log.d(TAG, "download speeds: MIN=" + DLmin + ", MED=" + DLmedian + ", MAX=" + DLmax + ", MEAN=" + DLmean);
+                            Log.d(TAG, "upload speeds: MIN=" + ULmin + ", MED=" + ULmedian + ", MAX=" + ULmax + ", MEAN=" + ULmean);
+                            stopParsingThread();
+                        }
+                    });
+
+
+                    Log.wtf(TAG, "Starting parsing...");
```
Comment (RkShaRkz) [inline app/src/main/java/cloudcity/Iperf3Monitor.java:280]: eh...

File: app/src/main/java/cloudcity/Iperf3Monitor.java [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+            }
+        });
+    }
+
+    private void startParsingThread(Iperf3Parser iperf3Parser) {
+        Log.d(TAG, "--> startParsingThread()");
+        handler.post(new Runnable() {
+            @Override
+            public void run() {
+                Log.d(TAG, "startParsingThread()::parsingCycle!\tshouldStop: " + shouldStop.get());
```
Comment (RkShaRkz) [inline app/src/main/java/cloudcity/Iperf3Monitor.java:295]: verbose. Log.v

**Review by RkShaRkz — COMMENTED** [review]
LGTM actually, but do address these comments.

**Review by slavkof — CHANGES_REQUESTED** [review]
Ima materijala da se dotera

File: app/src/main/java/cloudcity/Iperf3Monitor.java [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+
+        if (instance.handlerThread != null) {
+            instance.handlerThread.quitSafely();
+            try {
+                instance.handlerThread.join();
+            } catch (InterruptedException e) {
+                Log.e(TAG, "Exception " + e + " happened during shutdown() while joining the handlerThread !!! ", e);
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (RkShaRkz) [reply app/src/main/java/cloudcity/Iperf3Monitor.java:117]: fair point, done. also, should be synchronized. resolving.

File: app/src/main/java/cloudcity/Iperf3Monitor.java (hunk 4 above)
Comment (RkShaRkz) [reply app/src/main/java/cloudcity/Iperf3Monitor.java:146]: done. resolving.

Comment (RkShaRkz) [reply app/src/main/java/cloudcity/Iperf3Monitor.java:146]: actually, will still throw if it happens.

File: app/src/main/java/cloudcity/Iperf3Monitor.java [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+                    String originalFilePath = latestIperf3RunResult.input.iperf3rawIperf3file;
+                    String targetFilePath = originalFilePath + ".tmp";
+
+                    try {
+                        copyFile(originalFilePath, targetFilePath);
+                    } catch (IOException e) {
+//                        throw new RuntimeException(e);
+                        Log.e(TAG, "Exception " + e + " happened during iperf3 raw file copy!", e);
+                    }
+
```
[the remark this answers is not in the document]
Comment (RkShaRkz) [reply app/src/main/java/cloudcity/Iperf3Monitor.java:149]: throwing RuntimeException above will do that, resolving.

File: app/src/main/java/cloudcity/Iperf3Monitor.java (hunk 5 above)
Comment (RkShaRkz) [reply app/src/main/java/cloudcity/Iperf3Monitor.java:222]: done, resolving.

File: app/src/main/java/cloudcity/Iperf3Monitor.java (hunk 6 above)
Comment (RkShaRkz) [reply app/src/main/java/cloudcity/Iperf3Monitor.java:271]: done, normalized. resolving.

File: app/src/main/java/cloudcity/Iperf3Monitor.java (hunk 7 above)
Comment (RkShaRkz) [reply app/src/main/java/cloudcity/Iperf3Monitor.java:280]: done. resolving.

File: app/src/main/java/cloudcity/Iperf3Monitor.java (hunk 8 above)
Comment (RkShaRkz) [reply app/src/main/java/cloudcity/Iperf3Monitor.java:295]: done. resolving.

File: app/src/main/java/cloudcity/dataholders/MetricsPOJO.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    private final double DLmedian;
+    private final double DLmean;
+    private final double DLmax;
+    private final double DLlast;
+
+    private final double ULmin;
+    private final double ULmedian;
+    private final double ULmean;
+    private final double ULmax;
+    private final double ULlast;
```
[the remark this answers is not in the document]
Comment (RkShaRkz) [reply app/src/main/java/cloudcity/dataholders/MetricsPOJO.java:15]: nope, but will add class-level javadoc. resolving.

File: app/src/main/java/cloudcity/dataholders/MetricsPOJO.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+        this.DLmedian = DLmedian;
+        this.DLmean = DLmean;
+        this.DLmax = DLmax;
+        this.DLlast = DLlast;
+        this.ULmin = ULmin;
+        this.ULmedian = ULmedian;
+        this.ULmean = ULmean;
+        this.ULmax = ULmax;
+        this.ULlast = ULlast;
+    }
```
[the remark this answers is not in the document]
Comment (RkShaRkz) [reply app/src/main/java/cloudcity/dataholders/MetricsPOJO.java:29]: nope, but will do something else. resolving.

File: app/src/main/java/cloudcity/dataholders/MetricsPOJO.java (hunk 2 above)
Comment (RkShaRkz) [reply app/src/main/java/cloudcity/dataholders/MetricsPOJO.java:62]: deleted. resolving.

File: app/src/main/java/de/fraunhofer/fokus/OpenMobileNetworkToolkit/Iperf3/Iperf3Fragment.java [hunk 13]
```diff
@@ -285,6 +288,24 @@ public void afterTextChanged(Editable s) {
...
+        });
```
[the remark this answers is not in the document]
Comment (RkShaRkz) [reply app/src/main/java/de/fraunhofer/fokus/OpenMobileNetworkToolkit/Iperf3/Iperf3Fragment.java:307]: done. but not in this file. and no try/catch either. resolving.

File: app/src/main/java/de/fraunhofer/fokus/OpenMobileNetworkToolkit/Iperf3/Iperf3Parser.java [hunk 14]
```diff
@@ -15,20 +17,35 @@
 
 
 public class Iperf3Parser {
+    private static final String TAG = Iperf3Parser.class.getSimpleName();
+
+    public static final Object END_MARKER = new Object();
```
Comment (RkShaRkz) [inline app/src/main/java/de/fraunhofer/fokus/OpenMobileNetworkToolkit/Iperf3/Iperf3Parser.java:22]: get rid of this.

Comment (RkShaRkz) [reply app/src/main/java/de/fraunhofer/fokus/OpenMobileNetworkToolkit/Iperf3/Iperf3Parser.java:22]: or you know what, just let it sit for a while, and when it causes conflicts then delete it. it's just used as a marker value to fire the "end" property change so we see when the parsing is over.

File: app/src/main/java/de/fraunhofer/fokus/OpenMobileNetworkToolkit/Iperf3/Iperf3Parser.java [hunk 15]
```diff
@@ -15,20 +17,35 @@
...
+    }
+
     Iperf3Parser(String pathToFile) {
         this.pathToFile = pathToFile;
         this.file = new File(this.pathToFile);
         try {
             br = new BufferedReader(new FileReader(file));
         } catch (FileNotFoundException ex) {
             System.out.println("File not found");
+            Log.e(TAG, "File not found!!! Path to file: "+pathToFile);
```
[the remark this answers is not in the document]
Comment (RkShaRkz) [reply app/src/main/java/de/fraunhofer/fokus/OpenMobileNetworkToolkit/Iperf3/Iperf3Parser.java:48]: not necessary, we don't want to change their original code too much. lets see if it's an issue first.

wont change, resolving.

File: app/src/main/java/de/fraunhofer/fokus/OpenMobileNetworkToolkit/Iperf3/Iperf3Parser.java [hunk 16]
```diff
@@ -42,6 +59,7 @@ public void parse(){
                 String event = obj.getString("event");
                 switch (event) {
                     case "start":
+                        Log.v(TAG, "Encountered START");
```
[the remark this answers is not in the document]
Comment (RkShaRkz) [reply app/src/main/java/de/fraunhofer/fokus/OpenMobileNetworkToolkit/Iperf3/Iperf3Parser.java:62]: nope, but instead od silently swallowing, we'll also log it as an error.

resolving.

File: app/src/main/java/de/fraunhofer/fokus/OpenMobileNetworkToolkit/Iperf3/Iperf3ResultsDataBase.java [hunk 17]
```diff
@@ -21,7 +21,7 @@
 public abstract class Iperf3ResultsDataBase extends RoomDatabase {
     private static volatile Iperf3ResultsDataBase INSTANCE;
 
-    static Iperf3ResultsDataBase getDatabase(final Context context) {
+    public static Iperf3ResultsDataBase getDatabase(final Context context) {
```
[the remark this answers is not in the document]
Comment (RkShaRkz) [reply app/src/main/java/de/fraunhofer/fokus/OpenMobileNetworkToolkit/Iperf3/Iperf3ResultsDataBase.java:24]: nope, won't do any of this. this file was changed only so we can instantiate it from outside of the package.
resolving.

**Review by RkShaRkz — COMMENTED** [review]
LGTM, but take another peek after commiting the CR commit since I was addressing them as I was going through this.

**Review by slavkof — APPROVED** [review]

**Pull request merged**